sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CommonCoderTest.java

```diff
@@ -0,0 +1,219 @@
+          .add("coder", test.getCoder())
+          .add("nested", test.getNested())
+          .add("serialized", test.getSerialized())
+          .toString();
+      ret.add(new Object[] {test, testname});
+    }
+    return ret.build();
+  }
+
+  @Parameter(0)
+  public OneCoderTestSpec testSpec;
+
+  @Parameter(1)
+  public String ignoredTestName;
+
+  private static CommonCoderTestSpec parseSpec(String spec) throws IOException {
+    ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+    return mapper.readValue(spec, CommonCoderTestSpec.class);
+  }
+
+  private static void assertCoderIsKnown(CommonCoder coder) {
+    assertThat("not a known coder", coders.keySet(), hasItem(coder.getUrn()));
+    for (CommonCoder component : coder.getComponents()) {
+      assertCoderIsKnown(component);
+    }
+  }
+
+  /** Converts from JSON-auto-deserialized types into the proper Java types for the known coders. */
+  private static Object convertValue(Object value, CommonCoder coderSpec, Coder coder) {
+    switch (coderSpec.getUrn()) {
+      case "urn:beam:coders:bytes:0.1": {
+        return ((String) value).getBytes(StandardCharsets.ISO_8859_1);
+      }
+      case "urn:beam:coders:kv:0.1": {
+        Coder keyCoder = ((KvCoder) coder).getKeyCoder();
+        Coder valueCoder = ((KvCoder) coder).getValueCoder();
+        Map<String, Object> kvMap = (Map<String, Object>) value;
+        Object k = convertValue(kvMap.get("key"), coderSpec.getComponents().get(0), keyCoder);
+        Object v = convertValue(kvMap.get("value"), coderSpec.getComponents().get(1), valueCoder);
+        return KV.of(k, v);
+      }
+      case "urn:beam:coders:varint:0.1":
+        return ((Number) value).longValue();
+      default:
+        throw new IllegalStateException("Unknown coder URN: " + coderSpec.getUrn());
+    }
+  }
+
+  private static Coder<?> instantiateCoder(CommonCoder coder) {
+    List<Coder<?>> components = new LinkedList<>();
+    for (CommonCoder innerCoder : coder.getComponents()) {
+      components.add(instantiateCoder(innerCoder));
+    }
+    switch (coder.getUrn()) {
+      case "urn:beam:coders:bytes:0.1":
+        return ByteArrayCoder.of();
+      case "urn:beam:coders:kv:0.1":
+        return KvCoder.of(components.get(0), components.get(1));
+      case "urn:beam:coders:varint:0.1":
+        return VarLongCoder.of();
+      default:
+        throw new IllegalStateException("Unknown coder URN: " + coder.getUrn());
+    }
+  }
+
+  @Test
+  public void executeSingleTest() throws IOException {
+    assertCoderIsKnown(testSpec.getCoder());
+    Coder coder = instantiateCoder(testSpec.getCoder());
+    Object testValue = convertValue(testSpec.getValue(), testSpec.getCoder(), coder);
+    Context context = testSpec.getNested() ? Context.NESTED : Context.OUTER;
+    byte[] encoded = CoderUtils.encodeToByteArray(coder, testValue, context);
+    assertThat(testSpec.toString(), encoded, equalTo(testSpec.getSerialized()));
```

Comment:
For non-determanistic coders, we will only be able to check decoding. I'd rather explicitly mark to skip the ones this doesn't work on. 

Perhaps there's some deep equals we could use?